state/store.go

```diff
@@ -0,0 +1,177 @@
+package state
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path"
+	"path/filepath"
+	"sync"
+)
+
+var (
+	ErrNotFound      = errors.New("not found")
+	ErrAlreadyExists = errors.New("already exists")
+)
+
+// A simple key<->RequestedState store.
+type Store struct {
+	RootDir string
+	values  map[string]*RequestedState
+
+	sync.RWMutex
+}
+
+func NewStore(rootdir string) *Store {
+	return &Store{
+		RootDir: rootdir,
+		values:  make(map[string]*RequestedState),
+	}
+}
+
+// Initialize must be called before performing any operation on the store. It
+// will attempt to restore the data from disk.
+func (s *Store) Initialize() error {
+	s.Lock()
+	defer s.Unlock()
+
+	if err := os.MkdirAll(s.RootDir, 0700); err != nil && !os.IsNotExist(err) {
+		return err
+	}
+
+	if err := s.restore(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *Store) path(key string) string {
+	return path.Join(s.RootDir, key+".json")
+}
+
+func (s *Store) restore() error {
+	files, err := ioutil.ReadDir(s.RootDir)
+	if err != nil {
+		return err
+	}
+	for _, fileinfo := range files {
+		file := fileinfo.Name()
+
+		// Verify the file extension.
+		extension := filepath.Ext(file)
+		if extension != ".json" {
+			return fmt.Errorf("invalid file extension for filename %s (%s)", file, extension)
+		}
+
+		// Load the object back.
+		value, err := s.load(path.Join(s.RootDir, file))
```

Comment:
Should we stop the restore if we can't load a single file?
I wonder if there are situations a crash before writing the whole file can cause half writes to the disk, and then cause us never able to restore. 
I would consider log error but still able to use the cache.